src/TakeoffController.cpp

```diff
@@ -0,0 +1,191 @@
+    geometry_msgs::TransformStamped transform;
+    bool success = transform_wrapper_.getTransformAtTime(transform,
+                                                    "foot1",
+                                                    "map",
+                                                    time,
+                                                    update_timeout_);
+    if (!success) {
+        ROS_ERROR("Failed to get current transform in TakeoffController::update");
+        return false;
+    }
+
+    if(state_ == TakeoffState::RAMP) {
+        if(transform.transform.translation.z > max_takeoff_start_height_) {
+          recorded_hover_throttle_ = throttle_;
+          throttle_ = std::min(throttle_ + takeoff_throttle_bump_, 100.0);
+          state_ = TakeoffState::ASCEND;
+        }
+        else
+        {
+          throttle_ = std::min(throttle_ + ((time - last_update_time_).toSec() * takeoff_throttle_ramp_rate_), 100.0);
+        }
+    }
+    else if(state_ == TakeoffState::ASCEND)
+    {
+        // Continue to ascend when done return success and turn off
+        if(transform.transform.translation.z > takeoff_complete_height_)
+        {
+            state_ = TakeoffState::DONE;
+        }
+    }
+    else if(state_ == TakeoffState::DONE)
+    {
+        ROS_ERROR("Tried to update takeoff handler when in DONE state");
+        return false;
+    }
+    else
+    {
+        ROS_ASSERT_MSG(false, "Invalid state in takeoff controller");
+    }
+
+    // Fill in the uav_command's information
+    uav_command.header.stamp = time;
+
+    uav_command.throttle = throttle_;
+
+    // Check that none of the throttle values are infinite before returning
+    if (!std::isfinite(uav_command.throttle)
+     || !std::isfinite(uav_command.data.pitch)
+     || !std::isfinite(uav_command.data.roll)
+     || !std::isfinite(uav_command.data.yaw)) {
+        ROS_ERROR(
+            "Part of command is not finite in TakeoffHandler::update");
+        return false;
+    }
+
+    // Print the throttle information
+    ROS_DEBUG("Throttle: %f Pitch: %f Roll: %f Yaw: %f",
+             uav_command.throttle,
+             uav_command.data.pitch,
+             uav_command.data.roll,
+             uav_command.data.yaw);
+
+    last_update_time_ = time;
+    return true;
+}
+
+bool TakeoffController::waitUntilReady()
+{
+    geometry_msgs::TransformStamped transform;
+    bool success = transform_wrapper_.getTransformAtTime(transform,
+                                                    "foot1",
+                                                    "map",
+                                                    ros::Time(0),
+                                                    startup_timeout_);
+    if (!success)
+    {
+        ROS_ERROR("Failed to fetch initial transform");
+        return false;
+    }
+
+    // Mark the last update time the current time as we could not have
+    // received a message before this
+    last_update_time_ = ros::Time::now();
```

Comment:
Actually in quad velocity controller you just the get max updated time of the 3 interpolation classes. I'm going to leave this as ros::Time::now and leave it as a comment. The time is only used for the calculation of the ramping speed. I will leave a clarifying comment. Same goes for the LandPlanner.